src/pwm.rs

```diff
@@ -0,0 +1,186 @@
+    /// Using the returned device otherwise results in **undefined behavior**!
+    ///
+    /// Also note, that the given frequency is only to be used by [`embedded_hal::Pwm::get_period`] or [`PWMDevice::get_frequency`].
+    /// Just setting this to `x` will only result in those two functions returning `x` but will have no other impact, if
+    /// for example the given [`pwm_t`] is initialized by a board and the actually set frequency is unknown.
+    ///
+    pub unsafe fn new_without_init(dev: pwm_t, frequency: Hz) -> Self {
+        PWMDevice {
+            dev,
+            channel_duty_values: [0; CHANNELS],
+            frequency,
+        }
+    }
+
+    /// Initializes the [`PWMDevice`] with the given frequency in Hz and resolution of a period/duty cycle.
+    ///
+    /// If the given frequency and resolution values can not be acchieved on the current device, the
+    /// resolution will be kept the same while the frequency will be lowered until the device can handle the combination.
+    /// The resulting frequency is written into the [`PWMDevice`].
+    ///
+    /// Uses a [`Result`] in anticipation of a more usefull error handling in the future.
+    ///
+    /// Returns if the initialization was a success
+    fn init(&mut self, mode: PWMMode, frequency: Hz, resolution: u16) -> Result<(), ()> {
+        let err = unsafe { riot_sys::pwm_init(self.dev, mode.to_c(), frequency, resolution) };
+
+        match err {
+            0 => Err(()),
+            freq => {
+                // Set frequency in the device
+                self.frequency = freq;
+                Ok(())
+            }
+        }
+    }
+
+    /// Returns the number of available channels for this device
+    pub fn channels(&self) -> u8 {
+        unsafe { riot_sys::pwm_channels(self.dev).min(CHANNELS as u8) }
+    }
+
+    /// Stops PWM generation on this device
+    pub fn power_off(&mut self) {
+        unsafe { riot_sys::pwm_poweroff(self.dev) }
+    }
+
+    /// Resumes PWM generation after power_off on this device
+    pub fn power_on(&mut self) {
+        unsafe {
+            riot_sys::pwm_poweron(self.dev);
+        }
+    }
+
+    /// Sets the duty-cycle for the given channel
+    ///
+    /// value: `0: 0%, u16::MAX: 100%` duty_cycle
+    pub fn set(&mut self, channel: u8, value: u16) {
+        unsafe {
+            riot_sys::pwm_set(self.dev, channel, value);
+        }
+
+        let channel = channel as usize;
+        // Ignore if entry does not exists because
+        // only embedded_hal interface cares about those values
+        // and the implementation already checks for this
+        if channel < (CHANNELS) {
+            self.channel_duty_values[channel] = value;
+        }
+    }
+
+    /// Returns the used frequency of this [`PWMDevice`] in Hz
+    ///
+    pub fn get_frequency(&self) -> u32 {
+        self.frequency
+    }
+}
+
+pub type Seconds = f32;
```

Comment:
This will make some platforms unhappy if any code with this is actually emitted -- and honestly, I think it's weird that embedded-hal asks that something be there.

Given there are no constraints on the `Time` associated type, how would you feel about directly using the Hz type (possibly in an embedded-time incarnation)? It's not exactly as intended, but it does also express a time -- in a weird way, but it should still be usable for whatever the intended applications are.

Reply:
It would be a bit confusing. Another Idea might be to use `Duration`? One could just extract seconds from it and would be fine.
(Btw. What problems arise with such a type definition?)

Reply:
The problem is that many RIOT platforms don't have hardware floating point support, and where there isn't any, it's easy to pull in soft float support.

Other ideas ... no good ones off my head.

In my ideal world, the crate providing Hz (might be `fugit`?) would also have a trait like (say) FractionalSeconds that'd be cheap to convert from with Hz by just being returning { num: 1, denom: freq_in_hz } and implement all the conversions to fixed and floating point times from there. (My original suggestion was to do this in documentation only, with a fixed num of 1).

As of now, I'm not sure how to best go forward.

Other than for completion of the trait, are there requirements on this that you have from your use case?

Reply:
Ah o.k. thanks for the explanation. O.k. then `Duration` would not help. No i don't have any special requirements and was just thinking about what would be most intuitive to use.
Maybe we could use `fugit::Duration`. It is basically the same but actually expresses a duration instead of a rate.
We then could write 
```Rust
get_period(&self) -> Duration<u32, 1, RESOLUTION> {
    Duration.from_ticks(RESOLUTION / self.frequency)
}
```
This way we allow the frequency to change during runtime. The division is not very nice though. (Maybe do even more fancy stuff with correct rounding?)

Reply:
That'd require RESOLUTION to be part of the PwmTimer type. Doable, but it'd need to have a good default that doesn't get in the way.

(I've asked around [on the Rust Embedded](https://matrix.to/#/!BHcierreUuwCMxVqOf:matrix.org/$hbOJas5l5iq7Op1Q6IHn6lqLMcLPic8ODJSpSf_ASv8?via=matrix.org&via=psion.agg.io&via=tchncs.de) channel, but unless something comes back there, your plan sounds good to me.).

Reply:
The feedback I got was that Hz would be an OK type for the duration -- it doesn't have the exact arithmetic properties, but what it expresses is still good, and users can convert it to a Duration of whichever resolution they work with easily.

(It was also pointed out that these traits are gone from 1.0 embedded-HAL, which is unfortunate but something we'll need to deal with on riot-wrappers as a whole anyway).

Reply:
By the way, we may not want to spend endless brain power here: The successor in 1.0 embedded HAL will not have any of that time stuff any more, and deal purely in duty cycles (at the current state of https://github.com/rust-embedded/embedded-hal/pull/430)

Reply:
That also seems to get rid of the `get_duty` thing which will save a lot of pain from this interface implementation. As for how to continue: Since the embedded-HAL people said it is o.k. to use Hz, we just use that and return our frequency for the time being?

Reply:
I think that'll be a good choice, yes.

Reply:
This was done in https://github.com/RIOT-OS/rust-riot-wrappers/pull/38/commits/c94fc52eb00db061f4fa12b1c2c8ac7e1fef9240